platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/CountUpLatch.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.common.threading;
+
+import static com.swirlds.common.utility.CompareTo.isLessThan;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.concurrent.Phaser;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Similar to a {@link java.util.concurrent.CountDownLatch}, but counts up instead.
+ */
+public class CountUpLatch {
```

Comment:
The more I see of what's being done, the more I think phaser may not be ideal.  I think what's really needed here is a Condition.  The methods are very similar, but it's intended for "wait until condition is true" rather than "wait for multiple threads to align".

For a Condition:
On construct, declare the underlying ReentrantLock, and obtain the Condition via [Lock.newCondition()](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/locks/Lock.html#newCondition()).
For Add/Set, change the AtomicLong value, then call [Condition.signalAll()](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/locks/Condition.html#signalAll()).
for await, call [Condition.await()](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/locks/Condition.html#await()) or [Condition.await(time, timeUnit)](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/locks/Condition.html#await(long,java.util.concurrent.TimeUnit))
Much closer to what you already have here, and easier to make fit the encapsulated structure you want to design.

Reply:
I no longer use the phaser